### PR TITLE
feat(circuits): remove circomlib fork in favour of latest and zk-kit

### LINKED ISSUE #772: Poseidon encryption/decryption support
Currently, we are using a [fork](https://github.com/weijiekoh/circomlib) of circomlib which offers the utilities for poseidon encryption and decryption. This is a core component of MACI, however this results in us using generally outdated circomlib(js) code. 

Unfortunately this [PR](https://github.com/iden3/circomlib/pull/60) was never merged (and should actually be moved to circomlibjs), thus we need to think of a more future-proof way of using this code while having access to the most up to date circomlib/circomlibjs code. 

Further issue: seems like when changing from BigInt to bigint in packages importing this fork, certain actions fail with this error: "Cannot find module 'big-integer' from '../domainobjs/node_modules/circomlibjs/src/smt_hashes_mimc.js'"

Options:

* add encryption/decryption code to MACI repo
* fork circomlibjs and submit PR again (or ask the original author as that's their code)
* keep using this fork only for poseidon encryption and use instead the latest version of circomlibjs for all other needs (definitely simpler option)

### PR DESCRIPTION
**Pull request by ctrlc03**
# Description

Remove circomlib fork for poseidon decryption circuit, and use zk-kit instead. Also update circomlib to the latest version and use circomkit for circuits management and testing

## Related issue(s)

fix #772

## Confirmation

- [x] I have read and understand MACI's [contributor guidelines](https://github.com/privacy-scaling-explorations/maci/blob/dev/CONTRIBUTING.md) and [code of conduct](https://github.com/privacy-scaling-explorations/maci/blob/dev/CODE_OF_CONDUCT.md).
- [x] I have read and understand MACI's [GitHub processes](https://github.com/privacy-scaling-explorations/maci/discussions/847).
- [x] I have read and understand MACI's [testing guide](https://maci.pse.dev/docs/testing).

### CODE REVIEW

File: circuits/ts/__tests__/utils/utils.ts [hunk 1]
```diff
@@ -1,4 +1,16 @@
-import type { WasmTester } from "circom_tester";
+import { Circomkit, type WitnessTester, type CircomkitConfig } from "circomkit";
+
+import fs from "fs";
+import path from "path";
+
+const configFilePath = path.resolve(__dirname, "..", "..", "..", "circomkit.json");
+const config = JSON.parse(fs.readFileSync(configFilePath, "utf-8")) as CircomkitConfig;
+
+// eslint-disable-next-line import/prefer-default-export
```
Comment (0xmad) [inline circuits/ts/__tests__/utils/utils.ts:9]: Why do you need this rule? It should be disabled in linter.

**Review by 0xmad — CHANGES_REQUESTED** [review]
@ctrlc03 thanks, just few comments.
Regarding input types: I mentioned it only for one case by it's related to all the files in the PR.

File: circuits/circom/poseidon/poseidon.circom [hunk 2]
```diff
@@ -0,0 +1,16 @@
+pragma circom 2.0.0;
+
+// https://github.com/weijiekoh/circomlib/blob/feat/poseidon-encryption/circuits/poseidon.circom
+include "./poseidon-cipher.circom";
+
+template Poseidon_OLD(nInputs) {
```
[the remark this answers is not in the document]
Comment (ctrlc03) [reply circuits/circom/poseidon/poseidon.circom:6]: it would conflict with circomlib's one if I didn't change the name (and yes technically is the old poseidon implementation which is what we have been using until now and won't change because we would need different zkeys)

File: circuits/ts/__tests__/utils/utils.ts (hunk 1 above)
Comment (ctrlc03) [reply circuits/ts/__tests__/utils/utils.ts:9]: oh yea that was copy paste from another repo, will remove

File: circuits/ts/__tests__/CeremonyParams.test.ts [hunk 3]
```diff
@@ -150,11 +182,34 @@ describe("Ceremony param tests", () => {
...
         const currentBallotRoot = ballotTree.root;
 
-        const generatedInputs = poll.processMessages(pollId);
+        const inputs = poll.processMessages(pollId);
 
         // Calculate the witness
-        const witness = await circuit.calculateWitness(generatedInputs, true);
-        await circuit.checkConstraints(witness);
+        const correctedInputs = {
+          inputHash: inputs.inputHash as unknown as bigint,
```
[the remark this answers is not in the document]
Comment (ctrlc03) [reply circuits/ts/__tests__/CeremonyParams.test.ts:189]: fixed

File: circuits/ts/__tests__/CeremonyParams.test.ts [hunk 4]
```diff
@@ -172,30 +206,56 @@ describe("Ceremony param tests", () => {
...
-          pollEndTimestamp: generatedInputs.pollEndTimestamp,
-        });
-
-        const hasherWitness = await hasherCircuit.calculateWitness(hasherCircuitInputs, true);
-        await hasherCircuit.checkConstraints(hasherWitness);
+          coordPubKey: inputs.coordPubKey as unknown as [bigint, bigint],
+          msgRoot: inputs.msgRoot as unknown as bigint,
+          currentSbCommitment: inputs.currentSbCommitment as unknown as bigint,
+          newSbCommitment: inputs.newSbCommitment as unknown as bigint,
+          pollEndTimestamp: inputs.pollEndTimestamp as unknown as bigint,
```
Comment (0xmad) [inline circuits/ts/__tests__/CeremonyParams.test.ts:215]: Here you don't need casts anymore.

File: circuits/ts/__tests__/ProcessMessages.test.ts [hunk 5]
```diff
@@ -148,19 +182,19 @@ describe("ProcessMessage circuit", function test() {
...
-        pollEndTimestamp: generatedInputs.pollEndTimestamp,
-      });
-
-      const hasherWitness = await hasherCircuit.calculateWitness(hasherCircuitInputs, true);
-      await circuit.checkConstraints(witness);
+        coordPubKey: inputs.coordPubKey as unknown as [bigint, bigint],
+        msgRoot: inputs.msgRoot as unknown as bigint,
+        currentSbCommitment: inputs.currentSbCommitment as unknown as bigint,
+        newSbCommitment: inputs.newSbCommitment as unknown as bigint,
+        pollEndTimestamp: inputs.pollEndTimestamp as unknown as bigint,
```
Comment (0xmad) [inline circuits/ts/__tests__/ProcessMessages.test.ts:191]: Here you can also remove casts

File: circuits/ts/__tests__/CeremonyParams.test.ts (hunk 4 above)
Comment (ctrlc03) [reply circuits/ts/__tests__/CeremonyParams.test.ts:215]: great catch thank you

**Review by 0xmad — APPROVED** [review]
@ctrlc03 thanks!

File: circuits/scripts/build_intel.sh
Comment (0xmad) [inline circuits/scripts/build_intel.sh:1]: Let's add new line at the end

File: circuits/circom/poseidon/poseidon.circom (hunk 2 above)
[the remark this answers is not in the document]
Comment (samajammin) [reply circuits/circom/poseidon/poseidon.circom:6]: I see, thanks for the explanation @ctrlc03. Should we create an issue to upgrade the implementation we're using (i.e. to use circomlib's)?

File: circuits/package.json [hunk 6]
```diff
@@ -27,11 +27,13 @@
...
   "dependencies": {
-    "circomlib": "https://github.com/weijiekoh/circomlib#ac85e82c1914d47789e2032fb11ceb2cfdd38a2b",
+    "@zk-kit/circuits": "^0.3.0",
+    "circomlib": "^2.0.5",
     "maci-core": "^1.1.2",
     "maci-crypto": "^1.1.2",
     "maci-domainobjs": "^1.1.2",
-    "snarkjs": "^0.7.2"
+    "snarkjs": "^0.7.2",
+    "circomkit": "^0.0.19"
```
[the remark this answers is not in the document]
Comment (samajammin) [reply circuits/package.json:36]: Seems like more moving pieces than is necessary, but maybe there was some requirement to do this I'm missing

File: cli/testScript.sh [hunk 7]
```diff
@@ -47,7 +47,7 @@ HARDHAT_CONFIG=./build/hardhat.config.js node build/ts/index.js genProofs \
     -w true 
 HARDHAT_CONFIG=./build/hardhat.config.js node build/ts/index.js proveOnChain \
     --poll-id 0 \
-    --proof-dir proofs/
+    --proof-dir proofs/ \
```
Comment (samajammin) [inline cli/testScript.sh:50]: ?

File: circuits/package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (ctrlc03) [reply circuits/package.json:36]: it was needed to get it working with zk-kit, as there they use import mapping and circom_tester does not allow directly to resolve mapped imports. Not much has changed on our side with this as the testing framework we were using, circom_tester, is what circomkit uses hunder the hood

File: cli/testScript.sh (hunk 7 above)
Comment (ctrlc03) [reply cli/testScript.sh:50]: that needed to be there for the next line to be considered part of the command, otherwise it would have been counted as a command of its own and result in a bash shell error

File: cli/ts/commands/proveOnChain.ts [hunk 8]
```diff
@@ -77,7 +77,11 @@ export const proveOnChain = async (
   const maciContractAddress = maciAddress || readContractAddress("MACI");
   const messageProcessorContractAddress = messageProcessorAddress || readContractAddress(`MessageProcessor-${pollId}`);
   const tallyContractAddress = tallyAddress || readContractAddress(`Tally-${pollId}`);
-  const subsidyContractAddress = subsidyAddress || readContractAddress(`Subsidy-${pollId}`);
+
+  let subsidyContractAddress;
```
Comment (samajammin) [inline cli/ts/commands/proveOnChain.ts:81]: These changes seem unrelated from the rest of this PR - are these updates fixing some bug?

File: cli/testScript.sh (hunk 7 above)
Comment (ctrlc03) [reply cli/testScript.sh:50]: basically something we missed on a previous commit

File: cli/ts/commands/proveOnChain.ts (hunk 8 above)
Comment (ctrlc03) [reply cli/ts/commands/proveOnChain.ts:81]: yup, needed that to make manual test work properly so included here, but can definitely add in a separate one

File: circuits/circom/ecdh.circom [hunk 9]
```diff
@@ -1,6 +1,7 @@
 pragma circom 2.0.0;
-include "../node_modules/circomlib/circuits/bitify.circom";
-include "../node_modules/circomlib/circuits/escalarmulany.circom";
+
+include "./bitify.circom";
```
Comment (samajammin) [inline circuits/circom/ecdh.circom:3]: How are we able to import this template like this now (vs. referencing circomlib from node_modules)?

Makes it seem like this template is within the current directory 🤔

File: circuits/circom/poseidon/poseidon.circom (hunk 2 above)
[the remark this answers is not in the document]
Comment (ctrlc03) [reply circuits/circom/poseidon/poseidon.circom:6]: I think it's something we should be tracking on zk-kit also, so that the poseidon-cipher circuits can be upgraded to use the new version (I've tried to implement with circomlib's new poseidon algorithm, but was unsuccessful in my first try)

File: circuits/circom/ecdh.circom (hunk 9 above)
Comment (ctrlc03) [reply circuits/circom/ecdh.circom:3]: https://github.com/privacy-scaling-explorations/maci/pull/1010#discussion_r1451600519 basically circom allows to map imports, and circomkit supports it directly via its json config file

File: circuits/circom/float.circom [hunk 10]
```diff
@@ -1,4 +1,5 @@
 pragma circom 2.0.0;
+
 include "../node_modules/circomlib/circuits/bitify.circom";
```
Comment (samajammin) [inline circuits/circom/float.circom:3]: But here it seems we import the bitify the old way still?

Comment (ctrlc03) [reply circuits/circom/float.circom:3]: must have missed this import then, good catch

File: circuits/circom/ecdh.circom (hunk 9 above)
Comment (samajammin) [reply circuits/circom/ecdh.circom:3]: I see - so should we update this across all imports so we do it consistently?

Comment (samajammin) [reply circuits/circom/ecdh.circom:3]: Maybe not a big deal but I think a downside of this approach is now it's less obvious where we're importing templates from (i.e. within our codebase vs. libraries)

File: circuits/circomkit.json [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+  "prime": "bn128",
+  "version": "2.1.5",
+  "circuits": "./circom/circuits.json",
+  "dirPtau": "./ptau",
+  "dirCircuits": "./circom",
+  "dirInputs": "./inputs",
+  "dirBuild": "./build",
+  "optimization": 2,
+  "inspect": true,
+  "include": ["./node_modules/circomlib/circuits", "./node_modules/@zk-kit/circuits/circom"],
```
Comment (samajammin) [inline circuits/circomkit.json:12]: So this is what allows us to import templates without specifying /node_modules/? i.e. https://github.com/privacy-scaling-explorations/maci/pull/1010#r1451601984

Comment (samajammin) [reply circuits/circomkit.json:12]: If so, what happens if circomlib & zk-kit have a template of the same name? How do we know which we're using?

Comment (ctrlc03) [reply circuits/circomkit.json:12]: zk-kit uses circomlib too, so it would never have a template with the same name. Same goes for every other circom project in existence I would think 🤔

Comment (ctrlc03) [reply circuits/circomkit.json:12]: but yea if we used 2 libraries with the same template file name that would be a problem, though I don't think is one we will ever have

File: circuits/circom/ecdh.circom (hunk 9 above)
Comment (ctrlc03) [reply circuits/circom/ecdh.circom:3]: I see what you mean. Dunno really, I quite like it as it's shorter but I definitely see your point. Can always put it back if the team feels like we should keep as is

Comment (ctrlc03) [reply circuits/circom/ecdh.circom:4]: I added comments throughout the templates so it is clear which templates are coming from circomlib/zk-kit vs internal

File: circuits/ts/__tests__/CalculateTotal.test.ts [hunk 12]
```diff
@@ -1,16 +1,16 @@
-import { expect } from "chai";
-import tester from "circom_tester";
+import { WitnessTester } from "circomkit";
 
-import path from "path";
-
-import { getSignal } from "./utils/utils";
+import { circomkit } from "./utils/utils";
```
Comment (kittybest) [inline circuits/ts/__tests__/CalculateTotal.test.ts:3]: The name is confusing to me hhh (I just found it's the instance instead of the package). Could we rename it? (if it's unnecessary, just ignore it)

File: circuits/ts/__tests__/CalculateTotal.test.ts [hunk 13]
```diff
@@ -1,16 +1,16 @@
-import { expect } from "chai";
-import tester from "circom_tester";
+import { WitnessTester } from "circomkit";
```
Comment (kittybest) [inline circuits/ts/__tests__/CalculateTotal.test.ts:1]: maybe we only need the type here? `import { type WitnessTester } from "circomkit"`

File: circuits/ts/__tests__/IncrementalQuinTree.test.ts [hunk 14]
```diff
@@ -15,56 +13,71 @@ describe("IncrementalQuinTree circuit", function test() {
...
-    circuitGeneratePathIndices = await tester.wasm(
-      path.resolve(__dirname, "../../circom/test", `quinGeneratePathIndices_test.circom`),
-    );
-    circuitQuinSelector = await tester.wasm(path.resolve(__dirname, "../../circom/test", `quinSelector_test.circom`));
-    splicerCircuit = await tester.wasm(path.resolve(__dirname, "../../circom/test", `splicer_test.circom`));
+    circuitLeafExists = await circomkit.WitnessTester("quinLeafExists", {
+      file: "./trees/incrementalQuinTree",
+      template: "QuinLeafExists",
+      params: [3],
+    });
```
Comment (kittybest) [inline circuits/ts/__tests__/IncrementalQuinTree.test.ts:26]: Just a question, I saw there is another circom called `incrementalMerkleTree`, and there was tests toward it before. Does it need to be tested?

File: circuits/ts/__tests__/MessageValidator.test.ts [hunk 15]
```diff
@@ -1,22 +1,60 @@
...
+    cmd: SignalValueType;
+    pubKey: SignalValueType;
+    sigR8: SignalValueType;
+    sigS: SignalValueType;
+    currentVoiceCreditBalance: SignalValueType;
+    currentVotesForOption: SignalValueType;
+    voteWeight: SignalValueType;
+    slTimestamp: SignalValueType;
+    pollEndTimestamp: SignalValueType;
+  }
```
Comment (kittybest) [inline circuits/ts/__tests__/MessageValidator.test.ts:28]: should we move this interface to **types.ts** file? since I saw the other interfaces were moved there.

File: cli/testScript.sh (hunk 7 above)
Comment (kittybest) [reply cli/testScript.sh:50]: my fault sorr TT

File: circuits/ts/__tests__/CalculateTotal.test.ts (hunk 12 above)
Comment (ctrlc03) [reply circuits/ts/__tests__/CalculateTotal.test.ts:3]: changed to circomkitInstance to be clearer

File: circuits/ts/__tests__/IncrementalQuinTree.test.ts (hunk 14 above)
Comment (ctrlc03) [reply circuits/ts/__tests__/IncrementalQuinTree.test.ts:26]: not sure I understand what you mean here, could you rephrase please

Comment (kittybest) [reply circuits/ts/__tests__/IncrementalQuinTree.test.ts]: I mean the `incrementalMerkleTree.circom` seems not be tested, will it be tested anywhere or we don’t need it?

Comment (ctrlc03) [reply circuits/ts/__tests__/IncrementalQuinTree.test.ts]: Oh right, yea this was not tested even before, and it's not even used actually. We always use the quinary one as that's what we use throughout the codebase

**Review by kittybest — APPROVED** [review]
